File: leetcode/solutions/0678.ValidParenthesisString/checkValidString.ts
export default function checkValidString(s: string): boolean {
  let openCount: number[] = [];
  let astCount: number[] = [];

  for (const [i, char] of [...s].entries()) {
    if (char === "(") {
      openCount.push(i);
    } else if (char === "*") {
      astCount.push(i);
    } else if (openCount.length > 0) {
      openCount.pop();
    } else if (astCount.length > 0) {
      astCount.pop();
    } else {
      return false;
    }
  }

  let i = openCount.length - 1;
  for (let j = astCount.length - 1; i >= 0 && j >= 0; i--, j--) {
    if (openCount[i] > astCount[j]) {
      return false;
    }
  }

  return i < 0;
}

File: leetcode/solutions/0678.ValidParenthesisString/checkValidString.go
package solution

func checkValidString(s string) bool {
	openCount := make([]int, 0, len(s))
	astCount := make([]int, 0, len(s))

	for i, r := range s {
		switch {
		case r == '(':
			openCount = append(openCount, i)
		case r == '*':
			astCount = append(astCount, i)
		case len(openCount) > 0:
			openCount = openCount[:len(openCount)-1]
		case len(astCount) > 0:
			astCount = astCount[:len(astCount)-1]
		default:
			return false
		}
	}

	i := len(openCount) - 1
	for j := len(astCount) - 1; i >= 0 && j >= 0; i, j = i-1, j-1 {
		if openCount[i] > astCount[j] {
			return false
		}
	}

	return i < 0
}
